[css-anchor-position-1] Implement parsing support for anchor-size()

tuankiet65 opened the pull request:
#### 348917fed580492c717b5176875bc301268f6c67
<pre>
[css-anchor-position-1] Implement parsing support for anchor-size()
<a href="https://rdar.apple.com/136444595">rdar://136444595</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=280148">https://bugs.webkit.org/show_bug.cgi?id=280148</a>

Reviewed by Antti Koivisto.

Implement parsing anchor-size() as a calc() function.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-fallback-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-invalid-fallback-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-parse-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-size-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-size-minmax-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-size-parse-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-size-replaced-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-transition-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-typed-om-expected.txt:
* Source/WebCore/css/CSSProperties.json: Extended grammar of several properties to accept anchor-size()
* Source/WebCore/css/CSSValueKeywords.in: Added new value keywords required by anchor-size()
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
(WebCore::CSSCalc::toCalculationValue): Added toCalculationValue() for AnchorSize
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
(WebCore::CSSCalc::evaluate): Added dummy evaluate() for AnchorSize
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
(WebCore::CSSCalc::checkDepth): renamed from checkDepthAndIndex.
Remove empty token range check, because anchor-size() is valid and allowed.
(WebCore::CSSCalc::isCalcFunction): Recognized anchor-size() as a valid calc() function.
(WebCore::CSSCalc::consumeAnchorSize): Added function to parse anchor-size()
(WebCore::CSSCalc::parseCalcFunction):
(WebCore::CSSCalc::parseCalcSum):
(WebCore::CSSCalc::parseCalcProduct):
(WebCore::CSSCalc::parseCalcValue):
(WebCore::CSSCalc::checkDepthAndIndex): Deleted.
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
(WebCore::CSSCalc::serializeMathFunctionArguments): Implemented serializeMathFunctionArguments for AnchorSize
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
(WebCore::CSSCalc::simplify): Added stub simplify() for AnchorSize
* Source/WebCore/css/calc/CSSCalcTree+Simplification.h:
* Source/WebCore/css/calc/CSSCalcTree.h: Added new AnchorSize type, representing an anchor-size()
* Source/WebCore/css/parser/CSSPropertyParserConsumer+LengthPercentage.cpp:
(WebCore::CSSPropertyParserHelpers::consumeLengthPercentage):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+LengthPercentage.h:
Added new argument to consumeLengthPercentage that controls parsing of anchor-size()
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeAutoOrLengthPercentage): Added new argument to control parsing of anchor-size()
(WebCore::CSSPropertyParserHelpers::consumeMarginSide): Allow parsing anchor-size()
(WebCore::CSSPropertyParserHelpers::consumeSide): Allow parsing anchor-size()
(WebCore::CSSPropertyParserHelpers::consumeInsetLogicalStartEnd): Allow parsing anchor-size()
* Source/WebCore/css/parser/CSSPropertyParserOptions.h:
Added a new parsing policy to control whether to allow anchor-size() or not.
* Source/WebCore/css/process-css-properties.py:
(ReferenceTerm): Added all optional parameters supported by &lt;length-percentage&gt;.

Canonical link: <a href="https://commits.webkit.org/284630@main">https://commits.webkit.org/284630@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fab5faa8c4314b1b43bd4fd5e353af93d3f00e29

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70000 "5 style errors") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22753 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74085 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21158 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72117 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57201 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21009 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Running apply-patch; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55538 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14023 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73066 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44995 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60370 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36019 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41660 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17798 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19535 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63582 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18149 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75800 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Compiled WebKit") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14225 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17374 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63237 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14261 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60445 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63176 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15537 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11194 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4806 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45204 "Built successfully") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46278 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47549 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46019 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

tuankiet65 commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -11286,6 +11286,14 @@
                 "url": "https://drafts.csswg.org/scroll-animations-1/#propdef-view-timeline-inset"
             }
+        },
+        "<anchor-size>": {
```
Not sure if I should define `<anchor-size>` like this, then use `CSSPropertyParsing::consumeAnchorSize`, or should I just use `CSSPropertyParserHelpers::consumeIdentRaw`

tuankiet65 replied:
Consensus is to use `CSSPropertyParserHelpers::consumeIdentRaw`

tuankiet65 commented on Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
```diff
@@ -84,8 +85,7 @@ struct ParserState {
 {
-    if (tokens.atEnd())
-        return ParseStatus::NoMoreTokens;
+    UNUSED_PARAM(tokens);
```
`anchor-size()` is a valid configuration, so it's possible the token range is empty. Would this affect correctness of anything?

tuankiet65 replied:
All tests are passing, so doesn't seem this would affect correctness of anything.

tuankiet65 commented on Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
```diff
@@ -702,6 +703,20 @@ static std::optional<TypedChild> consumeProgress(CSSParserTokenRange& tokens, in
     return TypedChild { makeChild(WTFMove(op), *outputType), *outputType };
 }
 
+static std::optional<TypedChild> consumeLengthPercentage(CSSParserTokenRange& tokens, int depth, ParserState& state)
```
The goal is `consumeAnchor` will use this function to parse the fallback value too (in a separate patch)

tuankiet65 replied:
Now it's using `consumeValueWithoutSimplifyingCalc`

tuankiet65 commented on Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
```diff
@@ -1325,6 +1325,13 @@ std::optional<Child> simplify(Anchor& anchor, const SimplificationOptions& optio
 
+std::optional<Child> simplify(AnchorSize& anchorSize, const SimplificationOptions& options)
+{
+    UNUSED_PARAM(anchorSize);
```
What's `simplify` supposed to do? and what's the difference between `simplify` and `evaluate`? (in the context of `anchor-size()`, which has nothing to simplify)

tuankiet65 replied:
Chatted offline with Antti to figure out what `simplify` should do.

tuankiet65 commented on Source/WebCore/css/calc/CSSCalcTree.h:
```diff
@@ -1204,6 +1224,8 @@ OP_TUPLE_LIKE_CONFORMANCE(Abs, 1);
 OP_TUPLE_LIKE_CONFORMANCE(Progress, 3);
 OP_TUPLE_LIKE_CONFORMANCE(Anchor, 0);
+// TODO: fix this.
+OP_TUPLE_LIKE_CONFORMANCE(AnchorSize, 0);
```
What's the tuple-like for?

tuankiet65 commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+LengthPercentage.cpp:
```diff
@@ -165,12 +165,13 @@ std::optional<CSS::LengthPercentageRaw> LengthPercentageKnownTokenTypeNumberCons
 // MARK: - Consumer functions
 
-RefPtr<CSSPrimitiveValue> consumeLengthPercentage(CSSParserTokenRange& range, const CSSParserContext& context, ValueRange valueRange, UnitlessQuirk unitless, UnitlessZeroQuirk unitlessZero, NegativePercentagePolicy negativePercentage, AnchorPolicy anchorPolicy)
+RefPtr<CSSPrimitiveValue> consumeLengthPercentage(CSSParserTokenRange& range, const CSSParserContext& context, ValueRange valueRange, UnitlessQuirk unitless, UnitlessZeroQuirk unitlessZero, NegativePercentagePolicy negativePercentage, AnchorPolicy anchorPolicy, AnchorSizePolicy anchorSizePolicy)
```
One more policy sigh. I don't particularly like it, but unfortunately the properties that accept `anchor-size()` is a superset of the properties that accept `anchor()` (i.e some property accepts `anchor-size()` but not `anchor()`), hence the additional policy.

tuankiet65 replied:
Antti said it's okay to have a policy here. It's high time we made `consumeLengthPercentage` take `CSSPropertyParseOptions` though.

tuankiet65 commented on Source/WebCore/css/process-css-properties.py:
```diff
@@ -1497,7 +1497,11 @@ class ReferenceTerm:
         BuiltinSchema.Entry("length-percentage", "consumeLengthPercentage",
             BuiltinSchema.OptionalParameter("value_range", values={"[0,inf]": "ValueRange::NonNegative"}, default="ValueRange::All"),
-            BuiltinSchema.OptionalParameter("unitless", values={"unitless-allowed": "UnitlessQuirk::Allow"}, default="UnitlessQuirk::Forbid")),
+            BuiltinSchema.OptionalParameter("unitless", values={"unitless-allowed": "UnitlessQuirk::Allow"}, default="UnitlessQuirk::Forbid"),
```
C++ doesn't allow named arguments, and `anchorSizePolicy` is at the very end of `consumeLengthPercentage`, hence why I had to add all of its arguments (even though nothing uses it)

anttijk commented on Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
```diff
@@ -399,6 +400,32 @@ void serializeMathFunctionArguments(StringBuilder& builder, const IndirectNode<A
+        SerializationState state { };
+        serializeCalculationTree(builder, *anchorSize->fallback, state);
+    }
+}
```
This could use `serializeWithoutOmittingPrefix`

anttijk commented on Source/WebCore/css/calc/CSSCalcTree.h:
```diff
@@ -1204,6 +1224,8 @@ OP_TUPLE_LIKE_CONFORMANCE(Abs, 1);
 OP_TUPLE_LIKE_CONFORMANCE(Sign, 1);
 OP_TUPLE_LIKE_CONFORMANCE(Progress, 3);
 OP_TUPLE_LIKE_CONFORMANCE(Anchor, 0);
+// TODO: fix this.
```
TODO really applies to Anchor too.

tuankiet65 replied:
I'll move it up and maybe create a bugzilla

tuankiet65 replied:
https://bugs.webkit.org/show_bug.cgi?id=280798

anttijk commented on Source/WebCore/css/process-css-properties.py:
```diff
@@ -4411,7 +4415,7 @@ def generate_call_string(self, *, range_string, context_string):
                 return f"{builtin.consume_function_name}({range_string}, {context_string}, {builtin.value_range}, {builtin.unitless})"
             elif isinstance(builtin, BuiltinLengthPercentageConsumer):
-                return f"{builtin.consume_function_name}({range_string}, {context_string}, {builtin.value_range}, {builtin.unitless})"
+                return f"{builtin.consume_function_name}({range_string}, {context_string}, {builtin.value_range}, {builtin.unitless}, {builtin.unitless_zero}, {builtin.negative_percentage}, {builtin.anchor}, {builtin.anchor_size})"
```
as discussed we should do this for anchor() too.

tuankiet65 replied:
I forgot that, will make a bugzilla.

tuankiet65 replied:
https://bugs.webkit.org/show_bug.cgi?id=280797

tuankiet65 committed:
[css-anchor-position-1] Implement parsing support for anchor-size()
rdar://136444595
https://bugs.webkit.org/show_bug.cgi?id=280148

Reviewed by Antti Koivisto.

Implement parsing anchor-size() as a calc() function.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-fallback-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-invalid-fallback-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-parse-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-size-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-size-minmax-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-size-parse-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-size-replaced-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-transition-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-typed-om-expected.txt:
* Source/WebCore/css/CSSProperties.json: Extended grammar of several properties to accept anchor-size()
* Source/WebCore/css/CSSValueKeywords.in: Added new value keywords required by anchor-size()
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
(WebCore::CSSCalc::toCalculationValue): Added toCalculationValue() for AnchorSize
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
(WebCore::CSSCalc::evaluate): Added dummy evaluate() for AnchorSize
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
(WebCore::CSSCalc::checkDepth): renamed from checkDepthAndIndex.
Remove empty token range check, because anchor-size() is valid and allowed.
(WebCore::CSSCalc::isCalcFunction): Recognized anchor-size() as a valid calc() function.
(WebCore::CSSCalc::consumeAnchorSize): Added function to parse anchor-size()
(WebCore::CSSCalc::parseCalcFunction):
(WebCore::CSSCalc::parseCalcSum):
(WebCore::CSSCalc::parseCalcProduct):
(WebCore::CSSCalc::parseCalcValue):
(WebCore::CSSCalc::checkDepthAndIndex): Deleted.
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
(WebCore::CSSCalc::serializeMathFunctionArguments): Implemented serializeMathFunctionArguments for AnchorSize
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
(WebCore::CSSCalc::simplify): Added stub simplify() for AnchorSize
* Source/WebCore/css/calc/CSSCalcTree+Simplification.h:
* Source/WebCore/css/calc/CSSCalcTree.h: Added new AnchorSize type, representing an anchor-size()
* Source/WebCore/css/parser/CSSPropertyParserConsumer+LengthPercentage.cpp:
(WebCore::CSSPropertyParserHelpers::consumeLengthPercentage):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+LengthPercentage.h:
Added new argument to consumeLengthPercentage that controls parsing of anchor-size()
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeAutoOrLengthPercentage): Added new argument to control parsing of anchor-size()
(WebCore::CSSPropertyParserHelpers::consumeMarginSide): Allow parsing anchor-size()
(WebCore::CSSPropertyParserHelpers::consumeSide): Allow parsing anchor-size()
(WebCore::CSSPropertyParserHelpers::consumeInsetLogicalStartEnd): Allow parsing anchor-size()
* Source/WebCore/css/parser/CSSPropertyParserOptions.h:
Added a new parsing policy to control whether to allow anchor-size() or not.
* Source/WebCore/css/process-css-properties.py:
(ReferenceTerm): Added all optional parameters supported by <length-percentage>.

Canonical link: https://commits.webkit.org/284630@main
